src/main/java/bwapi/BotWrapper.java

```diff
@@ -0,0 +1,191 @@
+            } finally {
+                frameBuffer.lockSize.unlock();
+                configuration.log("Main: onFrame asynchronous end");
+            }
+        } else {
+            configuration.log("Main: onFrame synchronous start");
+            handleEvents();
+            configuration.log("Main: onFrame synchronous end");
+        }
+
+    }
+
+    /**
+     * Allows an asynchronous bot time to finish operation
+     */
+    void endGame() {
+        if (botThread != null) {
+            try {
+                botThread.join();
+            } catch (InterruptedException ignored) {}
+        }
+    }
+
+    Throwable getLastBotThrow() {
+        lastBotThrowLock.lock();
+        Throwable output = lastBotThrow;
+        lastBotThrowLock.unlock();
+        return output;
+    }
+
+    private Thread createBotThread() {
+        return new Thread(() -> {
+            try {
+                configuration.log("Bot: Thread started");
+                while (!gameOver) {
+
+                    configuration.log("Bot: Ready for another frame");
+                    performanceMetrics.botIdle.time(() -> {
+                        frameBuffer.lockSize.lock();
+                        try {
+                            while (frameBuffer.empty()) {
+                                configuration.log("Bot: Waiting for a frame");
+                                frameBuffer.conditionSize.awaitUninterruptibly();
+                            }
+                        } finally {
+                            frameBuffer.lockSize.unlock();
+                        }
+                    });
+
+                    configuration.log("Bot: Peeking next frame");
+                    game.clientData().setBuffer(frameBuffer.peek());
+
+                    configuration.log("Bot: Handling frame #" + game.getFrameCount());
+                    handleEvents();
+
+                    configuration.log("Bot: Events done. Dequeuing frame #" + game.getFrameCount());
+                    frameBuffer.dequeue();
+                }
+            } catch (Throwable throwable) {
+                // Record the throw,
+                // Then allow the thread to terminate silently.
+                // The main thread will look for the stored throw.
+                lastBotThrowLock.lock();
+                lastBotThrow = throwable;
+                lastBotThrowLock.unlock();
+
+                // Awaken any threads waiting on bot progress
+                while (!frameBuffer.empty()) {
+                    frameBuffer.dequeue();
+                }
+            }
+        });
+    }
+
+    private void handleEvents() {
+        ClientData.GameData gameData = game.clientData().gameData();
+
+        // Populate gameOver before invoking event handlers (in case the bot throws)
+        for (int i = 0; i < gameData.getEventCount(); i++) {
+            gameOver = gameOver || gameData.getEvents(i).getType() == EventType.MatchEnd;
+        }
+
+        if (configuration.async) {

```

Comment:
for example if we keep `async` as a public variable, a user could change this on frame 10 or something, and I don't know if it's desirable to be able to swithc over to async on the fly (e.g. what side effects that could have)